sdk/src/main/scala/io/horizen/account/state/ProxyMsgProcessor.scala

```diff
@@ -0,0 +1,129 @@
+package io.horizen.account.state
+
+import io.horizen.account.abi.ABIUtil.{METHOD_ID_LENGTH, getABIMethodId, getArgumentsFromData, getFunctionSignature}
+import io.horizen.account.state.ProxyMsgProcessor._
+import io.horizen.account.state.events.ProxyInvocation
+import io.horizen.account.utils.WellKnownAddresses.PROXY_SMART_CONTRACT_ADDRESS
+import io.horizen.evm.Address
+import io.horizen.params.{MainNetParams, NetworkParams}
+import io.horizen.utils.BytesUtils
+import org.web3j.utils.Numeric
+import sparkz.crypto.hash.Keccak256
+
+trait ProxyProvider {
+
+}
+
+case class ProxyMsgProcessor(params: NetworkParams) extends NativeSmartContractMsgProcessor with ProxyProvider {
+
+  override val contractAddress: Address = PROXY_SMART_CONTRACT_ADDRESS
+  override val contractCode: Array[Byte] = Keccak256.hash("ProxySmartContractCode")
+
+  // TODO this must be updated when merging with 0.8.0, we should support consensusEpochNumber for hard fork activation
+  override def init(view: BaseAccountStateView): Unit = {
+    super.init(view)
+  }
+
+  def doInvokeSmartContractStaticCallCmd(invocation: Invocation, view: BaseAccountStateView, context: ExecutionContext): Array[Byte] = {
+    doInvokeSmartContractCmd(invocation, view, context, readOnly = true)
+  }
+
+  def doInvokeSmartContractCallCmd(invocation: Invocation, view: BaseAccountStateView, context: ExecutionContext): Array[Byte] = {
+    doInvokeSmartContractCmd(invocation, view, context, readOnly = false)
+  }
+
+  private def doInvokeSmartContractCmd(invocation: Invocation, view: BaseAccountStateView, context: ExecutionContext, readOnly : Boolean): Array[Byte] = {
+
+    val msg = context.msg
+    log.debug(s"Entering with invocation: $invocation")
+    // check that message contains a nonce, in the context of RPC calls the nonce might be missing
+    if (msg.getNonce == null) {
+      throw new ExecutionRevertedException("Call must include a nonce")
+    }
+
+    val value = invocation.value
+
+    // check that msg.value is greater or equal than zero
+    if (value.signum() < 0) {
+      throw new ExecutionRevertedException("Value must not be zero")
+    }
+
+    // check that sender account exists
+    if (!view.accountExists(invocation.caller)) {
+      throw new ExecutionRevertedException(s"Sender account does not exist: ${invocation.caller}")
+    }
+
+    val inputParams = getArgumentsFromData(invocation.input)
+
+    val cmdInput = InvokeSmartContractCmdInputDecoder.decode(inputParams)
+    val contractAddress = cmdInput.contractAddress
+    val data = cmdInput.dataStr
+
+    if (view.isEoaAccount(contractAddress)) {
+      throw new ExecutionRevertedException(s"smart contract address is an EOA")
+    }
+
+    val dataBytes = Numeric.hexStringToByteArray(data)
+    val additionalDepth = 0
+    val res = context.executeDepth(
```

Comment:
Why using executeDepth if additionalDepth == 0? Isn't the same as calling directly execute?

Reply:
Handled in https://github.com/HorizenOfficial/Sidechains-SDK/pull/911